File: C++/015_3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        sort(num.begin(), num.end());
        int N = (int)num.size();
        vector<vector<int> > res;
        res.clear();
        vector<int> now;
        for (int i = 0; i < N; i++) {
            if (i > 0 && num[i] == num[i - 1])
                continue;
            int j = i + 1, k = N - 1;
            int tar = -num[i];
            while (j < k) {
                if (num[j] + num[k] == tar) {
                    if (!(j > i + 1 && num[j] == num[j - 1]) && !(k < N - 1 && num[k] == num[k + 1])) {
                        now.clear();
                        now.push_back(num[i]);
                        now.push_back(num[j]);
                        now.push_back(num[k]);
                        res.push_back(now);
                    }
                    j++, k--;
                }
                else if (num[j] + num[k] < tar)
                    j++;
                else
                    k--;
            }
        }
        return res;
    }
};

File: Python/015_3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if nums is None or len(nums) < 3:
            return []
        nums.sort()
        res, i = [], 0
        while i < len(nums):
            j, k = i + 1, len(nums) - 1
            s = 0 - nums[i]
            while j < k:
                if nums[j] + nums[k] == s:
                    res.append([nums[i], nums[j], nums[k]])
                    while j + 1 < k and nums[j + 1] == nums[j]:
                        j += 1
                    j += 1
                    while j < k - 1 and nums[k - 1] == nums[k]:
                        k -= 1
                    k -= 1
                elif nums[j] + nums[k] < s:
                    while j + 1 < k and nums[j + 1] == nums[j]:
                        j += 1
                    j += 1
                else:
                    while j < k - 1 and nums[k - 1] == nums[k]:
                        k -= 1
                    k -= 1
            while i + 1 < len(nums) and nums[i + 1] == nums[i]:
                i += 1
            i += 1
        return res